File: backend/app/api/routes/activity.py
"""
API routes for user activities and analytics.
"""

from datetime import datetime
from typing import Dict, List, Optional

from fastapi import APIRouter, Depends, Header, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.database import get_db
from app.services.activity_service import ActivityService

router = APIRouter(prefix="/activity", tags=["Activity"])


# ============== Schemas ==============

class LogActivityRequest(BaseModel):
    action: str
    details: Optional[dict] = None


class ActivityResponse(BaseModel):
    id: str
    action: str
    details: Optional[dict] = None
    created_at: datetime
    
    class Config:
        from_attributes = True


class FocusSessionRequest(BaseModel):
    duration_minutes: int = 25
    break_duration_minutes: int = 5
    pomodoros_target: int = 4


class UpdateFocusSessionRequest(BaseModel):
    state: Optional[str] = None
    pomodoros_completed: Optional[int] = None
    memories_reviewed: Optional[int] = None
    memories_created: Optional[int] = None


class FocusSessionResponse(BaseModel):
    id: str
    duration_minutes: int
    break_duration_minutes: int
    pomodoros_target: int
    pomodoros_completed: int
    state: str
    memories_reviewed: int
    memories_created: int
    started_at: datetime
    ended_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# ============== Helper ==============

def get_user_id(x_user_id: Optional[str] = Header(None)) -> str:
    """Get user ID from header or use demo user."""
    return x_user_id or "demo-user"


# ============== Activity Routes ==============

@router.post("/log", response_model=ActivityResponse)
async def log_activity(
    request: LogActivityRequest,
    user_id: str = Depends(get_user_id),
    db: AsyncSession = Depends(get_db)
):
    """Log a user activity."""
    if db is None:
        raise HTTPException(status_code=503, detail="Database not configured")
    
    service = ActivityService(db)
    activity = await service.log_activity(user_id, request.action, request.details)
    return activity


@router.get("/recent", response_model=List[ActivityResponse])
async def get_recent_activities(
    limit: int = Query(default=50, le=200),
    action: Optional[str] = Query(default=None),
    user_id: str = Depends(get_user_id),
    db: AsyncSession = Depends(get_db)
):
    """Get recent activities for the current user."""
    if db is None:
        raise HTTPException(status_code=503, detail="Database not configured")
    
    service = ActivityService(db)
    activities = await service.get_recent_activities(user_id, limit, action)
    return activities


@router.get("/stats")
async def get_activity_stats(
    days: int = Query(default=30, le=365),
    user_id: str = Depends(get_user_id),
    db: AsyncSession = Depends(get_db)
) -> Dict[str, int]:
    """Get activity statistics for the current user."""
    if db is None:
        raise HTTPException(status_code=503, detail="Database not configured")
    
    service = ActivityService(db)
    return await service.get_activity_stats(user_id, days)


@router.get("/daily")
async def get_daily_activity(
    days: int = Query(default=7, le=90),
    user_id: str = Depends(get_user_id),
    db: AsyncSession = Depends(get_db)
) -> List[Dict]:
    """Get daily activity counts."""
    if db is None:
        raise HTTPException(status_code=503, detail="Database not configured")
    
    service = ActivityService(db)
    return await service.get_daily_activity(user_id, days)


# ============== Focus Session Routes ==============

@router.post("/focus/sessions", response_model=FocusSessionResponse)
async def create_focus_session(
    request: FocusSessionRequest,
    user_id: str = Depends(get_user_id),
    db: AsyncSession = Depends(get_db)
):
    """Create a new focus session."""
    if db is None:
        raise HTTPException(status_code=503, detail="Database not configured")
    
    service = ActivityService(db)
    session = await service.create_focus_session(
        user_id,
        request.duration_minutes,
        request.break_duration_minutes,
        request.pomodoros_target
    )
    return session


@router.get("/focus/sessions", response_model=List[FocusSessionResponse])
async def get_focus_sessions(
    limit: int = Query(default=20, le=100),
    include_active: bool = Query(default=True),
    user_id: str = Depends(get_user_id),
    db: AsyncSession = Depends(get_db)
):
    """Get focus sessions for the current user."""
    if db is None:
        raise HTTPException(status_code=503, detail="Database not configured")
    
    service = ActivityService(db)
    sessions = await service.get_focus_sessions(user_id, limit, include_active)
    return sessions


@router.patch("/focus/sessions/{session_id}", response_model=FocusSessionResponse)
async def update_focus_session(
    session_id: str,
    request: UpdateFocusSessionRequest,
    user_id: str = Depends(get_user_id),
    db: AsyncSession = Depends(get_db)
):
    """Update a focus session."""
    if db is None:
        raise HTTPException(status_code=503, detail="Database not configured")
    
    service = ActivityService(db)
    session = await service.update_focus_session(
        session_id, user_id,
        request.state,
        request.pomodoros_completed,
        request.memories_reviewed,
        request.memories_created
    )
    
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    
    return session


@router.post("/focus/sessions/{session_id}/end", response_model=FocusSessionResponse)
async def end_focus_session(
    session_id: str,
    user_id: str = Depends(get_user_id),
    db: AsyncSession = Depends(get_db)
):
    """End a focus session."""
    if db is None:
        raise HTTPException(status_code=503, detail="Database not configured")
    
    service = ActivityService(db)
    session = await service.end_focus_session(session_id, user_id)
    
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    
    return session


@router.get("/focus/stats")
async def get_focus_stats(
    days: int = Query(default=30, le=365),
    user_id: str = Depends(get_user_id),
    db: AsyncSession = Depends(get_db)
) -> Dict:
    """Get focus session statistics."""
    if db is None:
        raise HTTPException(status_code=503, detail="Database not configured")
    
    service = ActivityService(db)
    return await service.get_focus_stats(user_id, days)

File: frontend/src/lib/activity.ts
/**
 * Activity logging utility for tracking user actions in Neon database.
 */

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export type ActivityAction =
  | "search"
  | "upload"
  | "chat"
  | "focus_start"
  | "focus_complete"
  | "focus_pause"
  | "memory_view"
  | "memory_create"
  | "memory_update"
  | "memory_delete"
  | "insight_view"
  | "insight_dismiss"
  | "connection_explore"
  | "page_view"
  | "login"
  | "logout";

interface ActivityDetails {
  [key: string]: string | number | boolean | null | undefined;
}

/**
 * Log a user activity to the database.
 * This is fire-and-forget - errors are logged but don't affect the UI.
 */
export async function logActivity(
  action: ActivityAction,
  details?: ActivityDetails,
  userId?: string
): Promise<void> {
  try {
    await fetch(`${API_URL}/api/activity/log`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-User-Id": userId || "demo-user",
      },
      body: JSON.stringify({
        action,
        details: details || {},
      }),
    });
  } catch (error) {
    // Silent fail - activity logging shouldn't break the app
    console.debug("Activity log failed:", error);
  }
}

/**
 * Get recent activities for the current user.
 */
export async function getRecentActivities(
  limit: number = 50,
  action?: ActivityAction,
  userId?: string
): Promise<Activity[]> {
  try {
    const params = new URLSearchParams({ limit: limit.toString() });
    if (action) params.append("action", action);

    const response = await fetch(`${API_URL}/api/activity/recent?${params}`, {
      headers: {
        "X-User-Id": userId || "demo-user",
      },
    });

    if (response.ok) {
      return await response.json();
    }
  } catch (error) {
    console.debug("Failed to fetch activities:", error);
  }
  return [];
}

/**
 * Get activity statistics for the current user.
 */
export async function getActivityStats(
  days: number = 7,
  userId?: string
): Promise<ActivityStats | null> {
  try {
    const response = await fetch(`${API_URL}/api/activity/stats?days=${days}`, {
      headers: {
        "X-User-Id": userId || "demo-user",
      },
    });

    if (response.ok) {
      return await response.json();
    }
  } catch (error) {
    console.debug("Failed to fetch activity stats:", error);
  }
  return null;
}

// Types
export interface Activity {
  id: string;
  action: string;
  details: ActivityDetails | null;
  created_at: string;
}

export interface ActivityStats {
  total_activities: number;
  activities_by_action: Record<string, number>;
  daily_activities: Record<string, number>;
  most_active_hour: number;
  streak_days: number;
}

// Focus session helpers
export interface FocusSessionData {
  id: string;
  duration_minutes: number;
  break_duration_minutes: number;
  pomodoros_target: number;
  pomodoros_completed: number;
  state: string;
  memories_reviewed: number;
  memories_created: number;
  started_at: string;
  ended_at?: string;
}

export async function createFocusSession(
  data: {
    duration_minutes?: number;
    break_duration_minutes?: number;
    pomodoros_target?: number;
  },
  userId?: string
): Promise<FocusSessionData | null> {
  try {
    const response = await fetch(`${API_URL}/api/activity/focus`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-User-Id": userId || "demo-user",
      },
      body: JSON.stringify(data),
    });

    if (response.ok) {
      return await response.json();
    }
  } catch (error) {
    console.debug("Failed to create focus session:", error);
  }
  return null;
}

export async function updateFocusSession(
  sessionId: string,
  data: Partial<{
    state: string;
    pomodoros_completed: number;
    memories_reviewed: number;
    memories_created: number;
  }>,
  userId?: string
): Promise<FocusSessionData | null> {
  try {
    const response = await fetch(`${API_URL}/api/activity/focus/${sessionId}`, {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
        "X-User-Id": userId || "demo-user",
      },
      body: JSON.stringify(data),
    });

    if (response.ok) {
      return await response.json();
    }
  } catch (error) {
    console.debug("Failed to update focus session:", error);
  }
  return null;
}

export async function getActiveFocusSessions(
  userId?: string
): Promise<FocusSessionData[]> {
  try {
    const response = await fetch(`${API_URL}/api/activity/focus/active`, {
      headers: {
        "X-User-Id": userId || "demo-user",
      },
    });

    if (response.ok) {
      return await response.json();
    }
  } catch (error) {
    console.debug("Failed to fetch active focus sessions:", error);
  }
  return [];
}

// ============== Convenience Helper Functions ==============

/**
 * Log a search activity
 */
export function logSearch(
  query: string,
  mode: string,
  resultCount: number,
  userId?: string
): void {
  logActivity("search", { query, mode, result_count: resultCount }, userId);
}

/**
 * Log a file upload activity
 */
export function logUpload(
  filename: string,
  chunksCreated: number,
  userId?: string
): void {
  logActivity("upload", { filename, chunks_created: chunksCreated }, userId);
}

/**
 * Log a memory action (create, view, update, delete)
 */
export function logMemoryAction(
  action: "view" | "create" | "update" | "delete",
  memoryId: string,
  title?: string,
  userId?: string
): void {
  const activityAction = `memory_${action}` as ActivityAction;
  logActivity(activityAction, { memory_id: memoryId, title }, userId);
}

/**
 * Log a focus session action
 */
export function logFocusSession(
  action: "start" | "complete" | "pause",
  sessionType: string,
  durationMinutes: number,
  topic?: string,
  userId?: string
): void {
  const activityAction = `focus_${action}` as ActivityAction;
  logActivity(
    activityAction,
    {
      session_type: sessionType,
      duration_minutes: durationMinutes,
      topic,
    },
    userId
  );
}

/**
 * Log an insight interaction
 */
export function logInsightViewed(
  insightId: string,
  insightType: string,
  userId?: string
): void {
  logActivity("insight_view", { insight_id: insightId, type: insightType }, userId);
}

/**
 * Log a chat message
 */
export function logChat(
  sessionId: string,
  messageType: "user" | "assistant",
  userId?: string
): void {
  logActivity("chat", { session_id: sessionId, message_type: messageType }, userId);
}
